README.md
# markdown-it-github

> A [markdown-it] plugin generate github style markdown.

[markdown-it]: https://github.com/markdown-it/markdown-it

## Usage

> index.md

```markdown
# This is an H1
+ Red
+ Green
+ Blue
```

> index.js

```js
var fs = require('fs')
var plugin = require('markdown-it-github')
var md = require('markdown-it')().use(plugin)
var input = fs.readFileSync('index.md', 'utf-8')
var result = md.render(input)
console.log(result)
```

> Result

```html
<h1 class="md-github md-github__h1"><a class="md-github__anchor" name="this-is-an-h1" href="#this-is-an-h1"><svg class="md-github__octicon md-github__octicon-link" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16" version="1.1" width="16" height="16" aria-hidden="true"><path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path></svg></a>This is an H1</h1>
<ul class="md-github md-github__ul">
<li class="md-github md-github__li">Red</li>
<li class="md-github md-github__li">Green</li>
<li class="md-github md-github__li">Blue</li>
</ul>
```

### style

You can import [github.css](https://github.com/miaoxingsoldier/markdown-it-github/tree/master/lib/github.css) to replicates the github markdown style:

- add `github.css` stylesheet
- add `md-github__body` class to the container
- set a width for the container

```html
<!DOCTYPE html>
<html>
<head>
  <link rel="stylesheet" href="github.css">
  <style>
    .md-github__body {
      min-width: 200px;
      max-width: 980px;
    }
  </style> 
</head>
<body>
  <div class="md-github__body"><!-- markdown-it-github-output --></div>
</body>
</html>
```


CHANGELOG.md
<a name="0.5.0"></a>
# [0.5.0](https://github.com/miaoxingsoldier/markdown-it-github/compare/v0.4.0...v0.5.0) (2020-01-05)


### Features

* **style:** add github stylesheet ([bf8826b](https://github.com/miaoxingsoldier/markdown-it-github/commit/bf8826b))



<a name="0.4.0"></a>
# [0.4.0](https://github.com/miaoxingsoldier/markdown-it-github/compare/v0.3.0...v0.4.0) (2020-01-01)


### Features

* add image link ([c622c33](https://github.com/miaoxingsoldier/markdown-it-github/commit/c622c33))



<a name="0.3.0"></a>
# [0.3.0](https://github.com/miaoxingsoldier/markdown-it-github/compare/v0.2.0...v0.3.0) (2019-12-31)


### Features

* add task list ([e3312fc](https://github.com/miaoxingsoldier/markdown-it-github/commit/e3312fc))



<a name="0.2.0"></a>
# [0.2.0](https://github.com/miaoxingsoldier/markdown-it-github/compare/v0.1.0...v0.2.0) (2019-12-30)


### Features

* add heading anchor ([059dd20](https://github.com/miaoxingsoldier/markdown-it-github/commit/059dd20))



<a name="0.1.0"></a>
# [0.1.0](https://github.com/miaoxingsoldier/markdown-it-github/compare/5913dc3...v0.1.0) (2019-12-29)


### Features

* add markdown-it-github plugin ([5913dc3](https://github.com/miaoxingsoldier/markdown-it-github/commit/5913dc3))





test/utils.js
var path = require('path')
var fs = require('fs')

var utils = {
  getTestCase: function (testFile, theme) {
    var file = fs.readFileSync(path.resolve(__dirname, testFile), 'utf-8')
    var source = file.replace(/@theme/g, theme)
    var blocks = source.split(/\s@title/g)
    var tests = blocks.map(function (block, index) {
      var a = block.split('@output')
      var output = a[1].trim()
      var b = a[0].split('@input')
      var input = b[1].trim()
      var title = b[0].replace('@title ', '').trim()
      return {
        title: title,
        input: input,
        output: output
      }
    })
    return tests
  }
}

module.exports = utils


test/index.test.js
/* eslint-env jest */

var MarkdownIt = require('markdown-it')
var githubPlugin = require('../lib/github.js')
var utils = require('./utils.js')

describe('markdown-it-github', () => {
  var theme = 'md-github'
  var md = new MarkdownIt()
    .use(githubPlugin)

  var testcases = utils.getTestCase('index.txt', theme)
  testcases.forEach(testcase => {
    it(testcase.title, () => {
      var result = md.render(testcase.input)
      expect(result.trim()).toBe(testcase.output)
    })
  })
})


lib/task.js

var taskItemType = [
  'list_item_open',
  'paragraph_open',
  'inline'
].join('')

function isTaskItem (tokens, idx) {
  var tokenType = [
    tokens[idx].type,
    tokens[idx + 1].type,
    tokens[idx + 2].type
  ].join('')
  var taskMark = /^\[[ xX]\] /.test(tokens[idx + 2].content)
  return tokenType === taskItemType && taskMark
}

function getTokenCls (theme, type) {
  return [theme, type].join('__')
}

function getParentToken (tokens, idx) {
  var parentLevel = tokens[idx].level - 1
  for (var i = idx - 1; i >= 0; i--) {
    if (tokens[i].level === parentLevel) return tokens[i]
  }
  return null
}

var getTaskFunc = function (theme, Token) {
  var listCls = getTokenCls(theme, 'task-list')
  var itemCls = getTokenCls(theme, 'task-item')
  var checkboxCls = getTokenCls(theme, 'task-checkbox')

  function addCheckbox (tokens, idx) {
    var token = tokens[idx]
    token.children[0].content = token.children[0].content.slice(3)

    var checkboxToken = new Token('checkbox_input', 'input', 0)
    checkboxToken.attrSet('type', 'checkbox')
    checkboxToken.attrSet('disabled', 'true')
    checkboxToken.attrSet('class', checkboxCls)

    var checked = /^\[[xX]\]/.test(token.content)
    if (checked) {
      checkboxToken.attrSet('checked', 'true')
    }

    token.children.unshift(checkboxToken)
  }

  return function (tokens, idx) {
    var listToken = getParentToken(tokens, idx)
    if ((listToken.attrGet('class') || '').indexOf(listCls) === -1) {
      listToken.attrJoin('class', listCls)
    }

    var itemToken = tokens[idx]
    itemToken.attrJoin('class', itemCls)

    addCheckbox(tokens, idx + 2)
  }
}

module.exports = function taskPlugin (md, options = {}) {
  var theme = options.theme
  if (!theme) return

  function travalToken (state) {
    var addTask = getTaskFunc(theme, state.Token)

    var tokens = state.tokens
    tokens.forEach((token, idx) => {
      if (idx + 2 >= tokens.length || !isTaskItem(tokens, idx)) return
      addTask(tokens, idx)
    })
  }

  md.core.ruler.push('github_task', travalToken)
}


lib/image.js
var imageItemType = [
  'paragraph_open',
  'inline',
  'image',
  'paragraph_close'
].join('')

function isImageItem (tokens, idx) {
  var token = tokens[idx]
  if (!token.children || !token.children.length) return false

  var tokenType = [
    tokens[idx - 1].type,
    token.type,
    token.children[0].type,
    tokens[idx + 1].type
  ].join('')
  return tokenType === imageItemType
}

var getImageFunc = function (theme, Token) {
  return function addLink (tokens, idx) {
    var token = tokens[idx]

    var imageToken = token.children[0]
    imageToken.attrSet('style', 'max-width:100%;')

    var linkOpen = new Token('link_open', 'a', 1)
    linkOpen.attrSet('class', theme + '__a')
    linkOpen.attrSet('target', '_blank')
    linkOpen.attrSet('href', imageToken.attrGet('src'))

    var linkClose = new Token('link_close', 'a', -1)

    token.children.unshift(linkOpen)
    token.children.push(linkClose)
  }
}

module.exports = function imagePlugin (md, options = {}) {
  var theme = options.theme
  if (!theme) return

  function travalToken (state) {
    var addImage = getImageFunc(theme, state.Token)

    var tokens = state.tokens
    tokens.forEach((token, idx) => {
      if (idx < 1 || idx + 1 >= tokens.length || !isImageItem(tokens, idx)) return
      addImage(tokens, idx)
    })
  }

  md.core.ruler.push('github_image', travalToken)
}


lib/anchor.js
var GithubSlugger = require('github-slugger')
var slugger = new GithubSlugger()

var anchorSvg = '<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16" version="1.1" width="16" height="16" aria-hidden="true"><path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path></svg>'

var getAnchorCls = function (theme) {
  return [theme, 'anchor'].join('__')
}

var getAnchorIcon = function (theme) {
  const iconCls = ['octicon', 'octicon-link'].map(cls => [theme, cls].join('__')).join(' ')
  return anchorSvg.replace('<svg', '<svg class="' + iconCls + '"')
}

var getAnchorFunc = function (theme, Token) {
  var anchorCls = getAnchorCls(theme)
  var anchorIcon = getAnchorIcon(theme)

  return function (tokens, idx) {
    var children = tokens[idx + 1].children
    if (!children || !children.length) return

    const title = children
      .map(c => typeof c.content === 'string' ? c.content : '')
      .join('')
      .replace(/\s/g, '-')

    var anchor = slugger.slug(title)

    var anchorOpen = new Token('link_open', 'a', 1)
    anchorOpen.attrSet('class', anchorCls)
    anchorOpen.attrSet('name', anchor)
    anchorOpen.attrSet('href', '#' + anchor)

    var anchorContent = new Token('html_block', '', 0)
    anchorContent.content = anchorIcon

    var anchorClose = new Token('link_close', 'a', -1)

    children.unshift(anchorOpen, anchorContent, anchorClose)
  }
}

module.exports = function anchorPlugin (md, options = {}) {
  var theme = options.theme
  if (!theme) return

  function travalToken (state) {
    var addAnchor = getAnchorFunc(theme, state.Token)

    var tokens = state.tokens
    tokens.forEach((token, idx) => {
      if (token.type !== 'heading_open') return
      addAnchor(tokens, idx)
    })
  }

  md.core.ruler.push('github_anchor', travalToken)
}
